### ISSUE TITLE
[feature request] unique to work on nd tensors via dim arg

### ISSUE THREAD

**Issue by alicanb** [pytorch/pytorch#9997]
Can we have `unique` to return unique rows and cols as well, similar to `numpy.unique`? Example usage to give an idea:
```
>>> a = torch.tensor([[1, 0, 0], [1, 0, 0], [2, 3, 3]])
>>> torch.unique(a)
tensor([0, 1, 2, 3])
>>> torch.unique(a, dim=0)
tensor([[1, 0, 0], [2, 3, 3]])
>>> torch.unique(a, dim=1)
tensor([[1, 0], [1, 0], [2, 3]])
```

**Comment by ptrblck** [comment]
What would be the output for the second example if we change the input data:
```python
a = torch.tensor([[1, 0, 2], [1, 0, 0], [2, 3, 3]])
torch.unique(a, dim=1)
> tensor([[0, 1, 2], [0, 1], [2,3]])
```
This would create a `tensor` with different sizes in the dimensions.
[Numpy](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.unique.html) has an `axis` argument, but removes duplicated items along the provided axis, as shown in your first example.

**Comment by alicanb** [comment]
It should return some permutation of columns of the original tensor in your example, since all columns are unique.

**Comment by ptrblck** [comment]
Ok, got it. Sorry for the confusion.

**Comment by zou3519** [comment]
I have no plans to work on it @ptrblck and I don't think anyone else is planning on it (@alicanb?). 

That and the [cuda kernel for unique](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Unique.cu) look like good places to start.

**Comment by alicanb** [comment]
Not me. Feel free @ptrblck !

**Comment by vishwakftw** [comment]
This can be closed now I suppose.

**Comment by alicanb** [comment]
Yep, thanks @ptrblck

**Comment by NikhilKothari** [comment]
@ptrblck This doesn't work!

`a = torch.tensor([[1, 0, 2], [1, 0, 0], [2, 3, 3]])
  torch.unique(a, dim=1)

>> tensor([[0, 1, 2],
                  [0, 1, 0],
                  [3, 2, 3]])`

Please help!

**Comment by ptrblck** [comment]
@NikhilKothari The method works as intended and returns the unique columns, since you've specified `dim=1`, and reflects numpy's behavior:
```python
np.unique(a.numpy(), axis=1)
> array([[0, 1, 2],
         [0, 1, 0],
         [3, 2, 3]], dtype=int64)
```
If you want to get the unique elements from each row separately, you could use:
```python
[torch.unique(x) for x in a]
> [tensor([0, 1, 2]), tensor([0, 1]), tensor([2, 3])]
```
Note that the forum is a better place for these kind of questions ;)

**Comment by NikhilKothari** [comment]
@ptrblck Thanks! Actually I wanted to do this in a vectorized way without list comprehension. But I don't even know if it's possible. I mean is it possible in pytorch to have a tensor with varying row length? I've been looking for the answer since 3 days now.

And I am sorry, I'll take such questions in the forum going further.
Thanks for your help!

**Comment by ptrblck** [comment]
> But I don't even know if it's possible. I mean is it possible in pytorch to have a tensor with varying row length?

This might be possible with [NestedTensors](https://github.com/pytorch/nestedtensor) in the future, but I'm not familiar with the current roadmap.
